dlib/control/model_abstract.h

```diff
@@ -0,0 +1,169 @@
+// Copyright (C) 2017 Adrián Javaloy (adrian.javaloy@gmail.com)
+// License: Boost Software License   See LICENSE.txt for the full license.
+#undef DLIB_MODEL_ABSTRACT_Hh_
+#ifdef DLIB_MODEL_ABSTRACT_Hh_
+
+#include "approximate_linear_models_abstract.h"
+#include "../matrix.h"
+
+namespace dlib
+{
+
+    template <
+            template<typename, typename> class feature_extractor_type
+            >
+    class example_model
+    {
+        /*!
+            REQUIREMENTS ON feature_extractor
+                feature_extractor should implement the example_feature_extractor interface defined
+                at approximate_linear_models_abstract.h.
+
+            WHAT THIS OBJECT REPRESENTS
+                This is an example interface of a model class. This class represents an environment
+                where an agent will be deployed at. In other words, it is an interface between the
+                simulated/real world and the agent that has to be there. In short this class:
+                    - Holds information about the state, action and reward space.
+                    - Delegates the state representation to the feature_extractor.
+                    - Provides an initial state to start the agent.
+                    - Offers an interface to move in the world (look for actions, make steps in it
+                      and get a feedback/reward for them).
+        !*/
+    public:
+
+        // You have to define state, action and reward types.
+        typedef U state_type;
+        typedef V action_type;
+        typedef W reward_type;
+
+        // The feature extractor uses the same types as the model.
+        typedef feature_extractor_type<state_type, action_type> feature_extractor;
+
+        example_model(
+        );
+        /*!
+            ensures
+                - #get_feature_extractor() == feature_extractor()
+        !*/
+
+        action_type random_action(
+            const state_type &state
+        ) const;
+        /*!
+            ensures
+                - returns a random reachable action from state.
+        !*/
+
+        action_type find_best_action(
+            const state_type &state,
+            const matrix<double,0,1> &w
+        ) const;
+        /*!
+            requires
+                - w.size() == states_size()
+            ensures
+                - returns the action that maximizes the product
+                  dot(w, get_feature_extractor().get_features(state)).
+        !*/
+
+        const feature_extractor& get_feature_extractor(
+        ) const;
+        /*!
+            ensures
+                - returns the feature_extractor used by the model.
+        !*/
+
+        auto states_size(
+        ) const -> decltype(get_feature_extractor().num_features());
```

Comment:
Don't use decltype here.  It makes the documentation hard to read.  Should a user really have to go look up the feature extractor interface's num_features to find out that state_size will return an unsigned long?

Reply:
In general, don't use decltype in the documentation/abstract files unless it's needed.  Like if the return type is determined by some complex template output or something like that.  And even in those cases it's probably better to have a typedef in the class and to return that.

I guess my point is that the purpose of the stuff in the abstract file is to be documentation that's easily read by a user.  So writing it to be clear is the goal.  That's different from writing code where normally trying not to repeat yourself is a good idea, usually.  But in documentation, you don't want to write in exactly one place that something is an unsigned long and then constantly make people refer back to that to find out that yes, it's an unsigned long.